### PR TITLE
Adding implementation of C++23 ranges::to

### PR DESCRIPTION
**Pull request by lumberyard-employee-dm**
The [ranges::to](https://www.open-std.org/JTC1/SC22/WG21/docs/papers/2022/p1206r7.pdf) utility function allows conversion from a range to a container.

Added the C++23 ranges constructor and methods of `insert_range`, `append_range`, `prepend_range`, `assign_range` to the AZStd containers(`AZStd::list`, `AZStd::forward_list`, `AZStd::deque`, `AZStd::vector`, `AZStd::fixed_vector`, `AZStd::string`, `AZStd::fixed_string`, `AZStd::set`, `AZStd::multiset`, `AZStd::map`, `AZStd::multimap`, `AZStd::unordered_set`, `AZStd::unordered_multiset`, `AZStd::unordered_map`, `AZStd::unordered_multimap`) 

It allows such conversions such as `AZStd::string_view` -> `AZStd::vector<char>` or `AZStd::vector<AZStd::pair<int, AZStd::string>` <-> `AZStd::(unordered)_map<int, AZStd::string`

It also supports transforming ranges such as `AZStd::ranges::transform_view<FuncThatConvertsIntToString, AZStd::vector<int>` -> `AZStd::vector<AZStd::string>`

Adding implementation of C++20 `ranges::copy`, `ranges::copy_if`, `ranges::copy_n`, `ranges::copy_backward`, `ranges::move`, and `ranges::move_backward`.

Signed-off-by: lumberyard-employee-dm <56135373+lumberyard-employee-dm@users.noreply.github.com>

### CODE REVIEW

File: Code/Framework/AzCore/AzCore/std/ranges/ranges_algorithm.h [hunk 1]
```diff
@@ -1218,4 +1242,227 @@ namespace AZStd::ranges
...
+                bool_constant<weakly_incrementable<O>>,
+                bool_constant<indirectly_copyable<I, O>>
+                >, copy_result<I, O>>
+            {
+                for (; first != last; ++first, ++result)
+                {
+                    *result = *first;
+                }
+
+                return { AZStd::move(last), AZStd::move(result) };
```
Comment (hultonha) [inline Code/Framework/AzCore/AzCore/std/ranges/ranges_algorithm.h:1280]: question: You undoubtedly know better than me but I'm just curious is it definitely necessary to use `AZStd::move` here as part of the return?

File: Code/Framework/AzCore/AzCore/std/ranges/ranges_algorithm.h [hunk 2]
```diff
@@ -1218,4 +1242,227 @@ namespace AZStd::ranges
...
+                bool_constant<bidirectional_iterator<I1>>,
+                bool_constant<sentinel_for<S1, I1>>,
+                bool_constant<bidirectional_iterator<O>>,
+                bool_constant<indirectly_movable<I1, O>>
+                >, move_backward_result<I1, O>>
+            {
+                for (I1 iter{ last }; iter != first;)
+                {
+                    *--result = AZStd::ranges::iter_move(--iter);
+                }
```
Comment (hultonha) [inline Code/Framework/AzCore/AzCore/std/ranges/ranges_algorithm.h:1446]: question: Should this definitely be pre-decrement? 🤔 If we're pointing at the last element (and not `end`) then might be skip the last element? I could be getting things wrong I just wanted to check.

File: Code/Framework/AzCore/AzCore/std/ranges/ranges_to.h [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+    template<class C, class R, class... Args>
+    [[nodiscard]] constexpr auto to(R&& r, Args&&... args)
+        -> enable_if_t<conjunction_v<
+        bool_constant<input_range<R>>,
+        bool_constant<!view<C>>
+        >, C>
+    {
+        if constexpr (convertible_to<range_reference_t<R>, range_value_t<C>>)
+        {
+            // Container C contains supports a constructor that can accept the range directly
```
Comment (hultonha) [inline Code/Framework/AzCore/AzCore/std/ranges/ranges_to.h:184]: nit:

_before_
```
// Container C contains supports a constructor that can accept the range directly
```

_after_
```
// Container C supports a constructor that can accept the range directly
```
(repeated)

File: Code/Framework/AzCore/AzCore/std/ranges/ranges_to.h [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+                // Then use reserve to reserve the total number of elements needed
+                if constexpr (sized_range<R> && Internal::reservable_container<C>)
+                {
+                    c.reserve(ranges::size(r));
+                }
+
+                ranges::copy(r, Internal::container_inserter<range_reference_t<R>>(c));
+                return c;
+            }
+            // Otherwise the construct is ill-formed
```
Comment (hultonha) [inline Code/Framework/AzCore/AzCore/std/ranges/ranges_to.h:218]: nit: `construct` -> `constructor`

**Review by hultonha — COMMENTED** [review]
Overall changes look great I just have a couple of questions and some minor nits.

File: Code/Framework/AzCore/AzCore/std/iterator.h [hunk 5]
```diff
@@ -131,6 +131,32 @@ namespace AZStd
     using std::insert_iterator;
     using std::inserter;
 
+#if !defined(__cpp_lib_concepts)
+    // In order for pre C++20 back_inserter_iterator, front_inseter_iterator and insert_iterator
+    // to work with the range algorithms which require a weakly_incrementable iterator
+    // the difference_type type alias must not be void as it is in C++17
+    // https://en.cppreference.com/w/cpp/iterator/back_insert_iterator
+    // This is being workaround by specializing AZStd::iterator_traits for these types
+    // to provide a difference_type type alias
```
[the remark this answers is not in the document]
Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/AzCore/std/iterator.h:140]: That rewrite reads more clearly

File: Code/Framework/AzCore/AzCore/std/ranges/ranges_algorithm.h (hunk 2 above)
Comment (hultonha) [reply Code/Framework/AzCore/AzCore/std/ranges/ranges_algorithm.h:1446]: I see, interesting, thanks for the link :+1:

File: Code/Framework/AzCore/AzCore/std/ranges/ranges_algorithm.h (hunk 1 above)
Comment (hultonha) [reply Code/Framework/AzCore/AzCore/std/ranges/ranges_algorithm.h:1280]: Ah of course as they're passed in by value, makes sense, sorry should have realized that. Looks good! 🙂

**Review by hultonha — APPROVED** [review]
LGTM!

**Review by hultonha — APPROVED** [review]

**Review by byrcolin — APPROVED** [review]

**Pull request merged**